sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Distinct.java

```diff
@@ -78,26 +83,41 @@
    */
   public static <T, IdT> WithRepresentativeValues<T, IdT> withRepresentativeValueFn(
       SerializableFunction<T, IdT> fn) {
-    return new WithRepresentativeValues<T, IdT>(fn, null);
+    return new WithRepresentativeValues<>(fn, null);
   }
 
   @Override
   public PCollection<T> expand(PCollection<T> in) {
-    return in
-        .apply("CreateIndex", MapElements.via(new SimpleFunction<T, KV<T, Void>>() {
-          @Override
-          public KV<T, Void> apply(T element) {
-            return KV.of(element, (Void) null);
-          }
-        }))
-        .apply(Combine.<T, Void>perKey(
-            new SerializableFunction<Iterable<Void>, Void>() {
+    if (!(in.getWindowingStrategy().getWindowFn() instanceof NonMergingWindowFn)) {
+      LOG.warn("Distinct is not currently guaranteed to remove all duplicates when using merging"
+      + "window functions.");
+    }
+
+    PCollection<KV<T, Void>> combined =
+        in.apply("CreateIndex", MapElements.via(
+            new SimpleFunction<T, KV<T, Void>>() {
```

Comment:
This could be idiomatically `WithKeys.of(new SimpleFunction() { ... /* identity function */ ... }` but that is admittedly a triviality. I might call the node `"KeyByElement"`.